src/python-common/ceph/deployment/hostspec.py

```diff
@@ -13,6 +13,7 @@ def __init__(self,
                  addr=None,  # type: Optional[str]
                  labels=None,  # type: Optional[List[str]]
                  status=None,  # type: Optional[str]
+                 maintenance: bool = False,
```

Comment:
Seems strange to me to add an attribute "maintenance" that  is just a "status" value ...(imho)
I think that the host can be ok, offline, in_maintenance,  and more possibilities in the future.
Why if we do not have an attribute "offline", should we have an attribute "maintenance"?

Reply:
ok, what about adding an offline status?

Reply:
like so?

Reply:
Maybe i not explained well my "view" of the situation... i think that is not good to add more attributes/flags to the host ,, like "in-maintenance", off-line" ,... because in my mind these attributes are really different values of the attribute "status". 
What i propose is:
- set status as a list attribute in the host class: It has a fixed list of possible status, these status are in this moment 
`status = ["running", "offline", "in maintenance".]`
- add a method in the host class :
`def change_status(new_status):`
Because in this way we encapsulate the change of status in the class, and we can decide what change of status are  acceptable using a very simple "state machine" that will be easy to expend in the future and located in only one place. Example (probably we must not allow to change from "offline" to "in-maintenance)

With you proposal , we open the door to continue adding status as flags , and to move and spread the logic for the change of status along all the code.

Reply:
The problem is, changing `status` from `str` to `List[str]` only solves the problem partially. 

> Maybe i not explained well my "view" of the situation... i think that is not good to add more attributes/flags to the host ,, like "in-maintenance", off-line" ,... because in my mind these attributes are really different values of the attribute "status".

Right. Things like `drain` or `out-of-memory` or `disk-full` etc. yeah. I can see the issue. 

> What i propose is:
> 
> * set status as a list attribute in the host class: It has a fixed list of possible status, these status are in this moment
>   `status = ["running", "offline", "in maintenance".]`

My problem is, this is more or less the same thing as adding flags to the host, but will less type safety. Think about it, converting one data structure into the other format is trivial, thus we're only talking about the type of data structure here. 

> * add a method in the host class :
>   `def change_status(new_status):`
>   Because in this way we encapsulate the change of status in the class, and we can decide what change of status are  acceptable using a very simple "state machine" that will be easy to expend in the future and located in only one place. 

Handling the states is going to be implemented in `mgr/cephadm` anyway. I don't see the benefit right now. On the contrary, that doesn't work, as we have to prevent the `offline` status from being written to config-key right now, as it is only computed in-memory. Having a `def change_status(new_status: str):` method is going to make this irritating. 

> Example (probably we must not allow to change from "offline" to "in-maintenance)

Yeah, let's honestly try to not introduce any dependencies between the flags. 

> 
> With you proposal , we open the door to continue adding status as flags , and to move and spread the logic for the change of status along all the code.

Yes and no. It does open the door for adding new flags to `HostSpec`, but that has **no** impact on where the logic is going to be implemented.